tcmur_device.c

```diff
@@ -167,11 +170,30 @@ static void __tcmu_report_event(void *data)
 	 */
 	sleep(1);
 
-	pthread_mutex_lock(&rdev->state_lock);
+	pthread_mutex_lock(&rdev->rdev_lock);
+	/*
+	 * If the device is in recovery, will skip reporting the event
+	 * this time because the event should be report when the device
+	 * is reopening.
+	 */
+	if (rdev->flags & (TCMUR_DEV_FLAG_IN_RECOVERY |
+			   TCMUR_DEV_FLAG_STOPPING |
+			   TCMUR_DEV_FLAG_STOPPED)) {
+		pthread_mutex_unlock(&rdev->rdev_lock);
+		return;
+	}
+
+	rdev->flags |= TCMUR_DEV_FLAG_REPORTING_EVENT;
+	pthread_mutex_unlock(&rdev->rdev_lock);
```

Comment:
Why is `rdev_lock` released here?  `->report_event()` used to be called with it held.

Reply:
The `rdev_lock` should always be released when calling the handler's hooks. I think we need to pass a `has_lock` boolean parameter.